transport/src/main/java/io/netty/channel/AbstractChannel.java

```diff
@@ -446,6 +459,167 @@ protected final void readIfIsAutoRead() {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    public <T> Attribute<T> attr(AttributeKey<T> key) {
```

Comment:
I'd keep `AttributeMap`, add `Channel#attributeMap()` or something like that and deprecate `Channel#attr(...)` et al. That's just me; composition over inheritance and `has-a` instead of `is-a` relationships.

Reply:
@rkapsi that sounds like a reasonable option to me, but wouldn't it be more direct to think of `Channel` as having attributes rather than having a map which contains attributes? Maybe if `AttributeMap` was renamed to `AttributeHolder` or `AttributeAttachable` or something? This wouldn't prevent the implementation using composition if we didn't want to use inheritance.

Reply:
Oh yea, it (Channel) not being an extension of AttributeMap is already huge. Love it. In terms of test-ability and future use-cases it's still a bit PITA though. In order to test this you've to create a instance of Channel (my eye twitches) and if you wanted to implement some new type of Channel you've to actually extend AbstractChannel because it's the only thing that provides an implementation thereof (tears of sadness).